### DISCUSSION TITLE
2.15.0-RC

### DISCUSSION THREAD

**Discussion by Him188** [mamoe/mirai#2704] [Releases]
### 不兼容变更

- 删除 native 编译目标 (#2698)
- 弃用 `DeviceInfo` 构造器及 serializer, 以 `DeviceInfoBuilder` 和 `DeviceInfo.serializeToString` 替代 (#2674)
- `MessageSyncEvent` 的 `MessageSource` 的 `targetId` 由机器人 ID 变为实际的消息发送对象 (#2122, #2513)

### 新特性

- 更新登录协议 (#2613, #2592)
- 现在可以在 `BotAuthorization.authorize` 时通过 `BotAuthInfo` 获取导致重新登录的原因 (#2664)
- 优化插件 classpath 策略 (#2666)
  > [文档](https://github.com/mamoe/mirai/pull/2666/commits/f0fbd9a5dc792b2ef0453b7bc30cdb3db6c158a4)
    - 添加选项以允许插件不搜索其他插件的 classpath
    - 添加选项以允许插件阻止其他未依赖此插件的插件进行引用

- `AudioToSilkService` 现在稳定 (#2658)
- 支持使用 JSON 保存与读取 `PluginData` 与 `PluginConfig` (#2498 by @NoMathExpectation)
- 支持获取和管理群精华消息 (`Group.essences`) (#837, #2293, #2314 by @cssxsh)
- 以 Android Library 形式发布 Android 编译目标 (#2635, #2699 by @MrXiaoM)

### 优化和修复

- 修复扫码登录的已知问题 (无法进入登录验证等) (#2623, #2645, #2633 with @cssxsh)
- 修复刷新令牌后令牌不保存的问题 (#2590, #2584)
  > 这是一个很长时间的错误, 应该只在完全重启 mirai 或更新 bot 后才会影响
- 修复 `java.lang.IllegalStateException: wLoginSigInfoField is not yet initialized` (#1603, #2449, #2450 by @sandtechnology)

- 修复获取群活跃数据失败的问题 (#2609, #2610 by @cssxsh)
- 修复部分合并转发消息仅被作为APP消息接收的问题 (#2618, #2675)
- 修复 `MessageSource` 定位相关逻辑: (#2510, #2513)
    - 发送消息时实际发送的消息会忽略通过 MessagePreSendEvent 所进行的修改 (#2525)
    - 安卓客户端发送的 Quote 中 message source id 有误 (#2495)
    - OnlineMessageSourceFromGroupImpl 的 fromId 有误 (#2277)
- 修复获取群精华消息报错 (#2668, #2672 by @cssxsh)
- 修复群公告协议 (#2619 by @cssxsh)
- 修复 `OnlineAnnouncement.members` 只能获取 50 个人的问题 (#2531, #2530 by @cssxsh)
- 修复发送消息时实际发送的消息会忽略通过 `MessagePreSendEvent` 所进行的修改的问题 (#2525 by @NoMathExpectation)
- 优化文档 (#2639 by @cssxsh)
- 完善了用户文档部分的内容 (#2541 by @MikkoAyaka)
- 在文档注释手表协议不支持戳一戳事件 (#2572 by @7aGiven)
- 在消息日志中去除特殊控制符
- 添加 `QRCodeLoginListener.onCompleted()` 用于资源释放

## mirai-core-mock

- 实现 `Image.isUploaded` (#2582 by @Nambers)
- 修复 `mockHonorMember` (#2473 by @Nambers)

### 优化和修复

- 修复事件文档中 `AutoLoginEvent` 与 `StartupEvent` 的描述 (#2692 by @BakaFT)

### 新特性

- 新增一个勾选框，为用户提供是否使用阿里云镜像仓库的选项 (#2636 by @Samarium150)
- 添加 Gradle 插件的阿里云镜像仓库 (#1778, #2636 by @Samarium150)
- 支持设置 `jvmTarget` (#2387, #2631 by @Samarium150)
- 使插件在新建项目时创建 `gradle-wrapper.properties` 文件和 `src/test` 文件夹 (#2617, #2622 by @Samarium150)

<hr /><em>This discussion was created from the release <a href='https://github.com/mamoe/mirai/releases/tag/v2.15.0-RC'>2.15.0-RC</a>.</em>

**Comment by zhs141** [comment]
泰裤辣
